concat_idents can create empty identifiers

I posted this casually on the [tracking issue](https://github.com/rust-lang/rust/issues/29599) some months back:

> An amusing fact about the current implementation of concat_idents! is that it will accept an empty 
> argument list, making it possible to construct the empty string as an identifier:
>
> ```
> error[E0425]: cannot find value `` in this scope
>  --> src/main.rs:5:5
>   |
> 5 |     concat_idents!();
>   |     ^^^^^^^^^^^^^^^^^ not found in this scope
> ```
> (good luck actually doing anything with it)

I left it at that because I never thought there would be any serious discussion of stabilizing the feature.  Never say never, huh?

This is an easy fix and I am working on a PR.  I just made this issue so that I have a number to name the test case after. :duck:

**The Plot Thickens**

You can construct regular keywords as identifiers too:

```
error[E0425]: cannot find value `fn` in this scope
 --> src/main.rs:3:13
  |
3 |     let x = concat_idents!(fn);
  |             ^^^^^^^^^^^^^^^^^^ not found in this scope
```

but it does not appear that you can do Unforsakeable Things with this because you can't use the macro in patterns for some reason.

```rust
#![feature(concat_idents)]
fn main() {
    let concat_idents!(fn) = 3; // ERROR non-pattern macro in pattern position
    let x = concat_idents!(fn);
}
```

--- a/src/libcore/macros.rs
+++ b/src/libcore/macros.rs
@@ -606,8 +606,8 @@ mod builtin {
     #[macro_export]
     #[cfg(dox)]
     macro_rules! concat_idents {
-        ($($e:ident),*) => ({ /* compiler built-in */ });
-        ($($e:ident,)*) => ({ /* compiler built-in */ });
+        ($($e:ident),+) => ({ /* compiler built-in */ });
+        ($($e:ident,)+) => ({ /* compiler built-in */ });
     }
 
     /// Concatenates literals into a static string slice.

--- a/src/libstd/macros.rs
+++ b/src/libstd/macros.rs
@@ -450,8 +450,8 @@ pub mod builtin {
     #[unstable(feature = "concat_idents_macro", issue = "29599")]
     #[macro_export]
     macro_rules! concat_idents {
-        ($($e:ident),*) => ({ /* compiler built-in */ });
-        ($($e:ident,)*) => ({ /* compiler built-in */ });
+        ($($e:ident),+) => ({ /* compiler built-in */ });
+        ($($e:ident,)+) => ({ /* compiler built-in */ });
     }
 
     /// Concatenates literals into a static string slice.

--- a/src/libsyntax_ext/concat_idents.rs
+++ b/src/libsyntax_ext/concat_idents.rs
@@ -31,6 +31,11 @@ pub fn expand_syntax_ext<'cx>(cx: &'cx mut ExtCtxt,
         return base::DummyResult::expr(sp);
     }
 
+    if tts.is_empty() {
+        cx.span_err(sp, "concat_idents! takes 1 or more arguments.");
+        return DummyResult::expr(sp);
+    }
+
     let mut res_str = String::new();
     for (i, e) in tts.iter().enumerate() {
         if i & 1 == 1 {

--- a/src/test/ui/issue-50403.rs
+++ b/src/test/ui/issue-50403.rs
@@ -1,0 +1,15 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(concat_idents)]
+
+fn main() {
+    let x = concat_idents!(); //~ ERROR concat_idents! takes 1 or more arguments
+}

--- a/src/test/ui/issue-50403.stderr
+++ b/src/test/ui/issue-50403.stderr
@@ -1,0 +1,8 @@
+error: concat_idents! takes 1 or more arguments.
+  --> $DIR/issue-50403.rs:14:13
+   |
+LL |     let x = concat_idents!(); //~ ERROR concat_idents! takes 1 or more arguments
+   |             ^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+

